### PR TITLE
Ensure Background tasks name is visible for old entries

### LINKED ISSUE #13296: Background Tasks are not showing title for existing entries
### Is your feature request related to a problem? Please describe.
In a recent change to the background tasks, a title was added to the background tasks. This work great for entries that have not been saved into the database. However, for entries that that are already saved in the database, we do not have a title to show so it leads to no name being displayed on the UI.

image

### Describe the solution you'd like
Since title is not an editable field, we should always set the title to the default-settings when showing task to on the UI.

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #13296

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs [hunk 1]
```diff
@@ -270,12 +270,18 @@ private static BackgroundTaskSettings GetSettings(IBackgroundTask task, Backgrou
...
 
-            if (document.Settings.ContainsKey(name))
+            BackgroundTaskSettings settings;
+            var defaultSettings = task.GetDefaultSettings();
+
+            if (document.Settings.TryGetValue(name, out settings))
             {
-                return document.Settings[name];
+                // Set the name using the default settings to ensure it populates old documents.
+                settings.Name = defaultSettings.Name;
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:279]: Here it depends if the `BackgroundTaskDocument document` was retrieved by using `LoadDocumentAsync()` or by using `GetDocumentAsync()`, the first returns a mutable document that we use in the Post Edit, but the second returns an immutable document that should not be updated. So here we would need to make the distinction.


    /// <summary>
    /// Loads the background task document from the store for updating and that should not be cached.
    /// </summary>
    public Task<BackgroundTaskDocument> LoadDocumentAsync() => _documentManager.GetOrCreateMutableAsync();

    /// <summary>
    /// Gets the background task document from the cache for sharing and that should not be updated.
    /// </summary>
    public Task<BackgroundTaskDocument> GetDocumentAsync() => _documentManager.GetOrCreateImmutableAsync();

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:279]: @jtkech yes but it should not matter in this case. Both get requests "List and Edit" are being set just for UI purposes. The records are being read and not saved. The post request to the edit method is where we actually update the database and there we already use Load. This change will fix it wether we want to save or just show it.

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:279]: Even if here we only update an immutable string, we should not mutate a shared cached object.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:279]: Are you saying we should use Load method everywhere in this case?

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:279]: No, better to still use the Get method in the Get Edit.

Then would need to look at the code, but normally only set the view model with the desired name.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:279]: I'll look at the code more tomorrow

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs]: @jtkech I made some changes to the code.

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs]: Okay, looks good!

File: src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs [hunk 2]
```diff
@@ -130,16 +152,19 @@ public async Task<IActionResult> Edit(string name)
             }
 
             var document = await _backgroundTaskManager.GetDocumentAsync();
-
-            var settings = GetSettings(task, document);
+            var defaultSettiongs = task.GetDefaultSettings();
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:155]: `defaultSettiongs` => `defaultSettings`
Idem in other places

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs]: Fixed

**Review by sebastienros — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs [hunk 3]
```diff
@@ -63,22 +62,44 @@ public async Task<IActionResult> Index(AdminIndexOptions options, PagerParameter
...
+                }
+
+                return new BackgroundTaskEntry()
+                {
+                    Title = defaultSettings.Title,
+                    Name = defaultSettings.Name,
+                    Description = defaultSettings.Description,
+                    Enable = defaultSettings.Enable,
+                };
+            });
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:87]:
```csharp
var items = _backgroundTasks.Select(task =>
{
    var defaultSettings = task.GetDefaultSettings();

    var taskEntry = new BackgroundTaskEntry { Title = defaultSettings.Title };

    if (document.Settings.TryGetValue(task.GetTaskName(), out var settings))
    {
        taskEntry.Name = settings.Name;
        taskEntry.Description = settings.Description;
        taskEntry.Enable = settings.Enable;
    }
    else
    {
        taskEntry.Name = defaultSettings.Name;
        taskEntry.Description = defaultSettings.Description;
        taskEntry.Enable = defaultSettings.Enable;
    }
});
```

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:87]: May be you could assign the `settings` directly if the value exists on the document settings

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:87]: I don't understand the ask.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:87]: @hishamco `taskEntry` is of `BackgroundTaskEntry` type where as `settings` is of `BackgroundTaskSettings`

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Controllers/BackgroundTaskController.cs:87]: Leave my last comment, coz I saw you did something similar in your PR, anyhow the above snippet might much cleaner

**Pull request merged**